File: backend/app/services/simulator.py
import logging
from typing import List, Dict, Any
from motor.motor_asyncio import AsyncIOMotorDatabase
from bson import ObjectId

logger = logging.getLogger("app.services.simulator")

async def run_what_if_simulation(
    db: AsyncIOMotorDatabase,
    assessment_id: str,
    selected_intervention_ids: List[str]
) -> Dict[str, Any]:
    """
    Implements the what-if simulation calculation with the max-per-leak-point overlap rule (§6).
    Reads current_co2e as an immutable input from the calculation engine.
    """
    # 1. Fetch Assessment
    assessment = await db.assessments.find_one({"_id": assessment_id})
    if not assessment and ObjectId.is_valid(assessment_id):
        assessment = await db.assessments.find_one({"_id": ObjectId(assessment_id)})

    current_co2e = float(assessment.get("total_co2e", 0.0)) if assessment else 0.0

    if not selected_intervention_ids:
        return {
            "current_co2e": current_co2e,
            "projected_co2e": current_co2e,
            "reduction_abs": 0.0,
            "reduction_pct": 0.0,
            "investment": 0.0,
            "payback": 0,
            "roi": 0.0
        }

    # 2. Fetch Selected Recommendations / Interventions
    selected_recs = []
    for sel_id in selected_intervention_ids:
        # Search by recommendation _id or intervention_id
        rec = await db.recommendations.find_one({"_id": sel_id, "assessment_id": assessment_id})
        if not rec and ObjectId.is_valid(sel_id):
            rec = await db.recommendations.find_one({"_id": ObjectId(sel_id), "assessment_id": assessment_id})
        if not rec:
            rec = await db.recommendations.find_one({"intervention_id": sel_id, "assessment_id": assessment_id})
        
        if rec:
            selected_recs.append(rec)
        else:
            # Fallback lookup in intervention library directly
            int_doc = await db.intervention_library.find_one({"id": sel_id})
            if int_doc:
                selected_recs.append({
                    "intervention_id": sel_id,
                    "intervention": int_doc,
                    "estimated_cost_range": [int_doc.get("estimated_cost_min", 0), int_doc.get("estimated_cost_max", 0)],
                    "estimated_co2_reduction_range": [int_doc.get("estimated_co2_reduction_min", 0), int_doc.get("estimated_co2_reduction_max", 0)],
                    "payback_period_months": int_doc.get("payback_period_months", 24),
                    "applicable_leak_point": int_doc.get("applicable_leak_types", ["general"])[0] if int_doc.get("applicable_leak_types") else "general"
                })

    # 3. Apply Overlap Rule: Max CO2 Reduction per Leak Point
    leak_point_reduction_pcts = {}
    total_investment = 0.0
    weighted_payback_sum = 0.0
    roi_list = []

    for rec in selected_recs:
        intervention = rec.get("intervention", {})
        
        # Investment (midpoint of cost range)
        cost_range = rec.get("estimated_cost_range") or [
            intervention.get("estimated_cost_min", 0),
            intervention.get("estimated_cost_max", 0)
        ]
        cost_mid = (float(cost_range[0]) + float(cost_range[1])) / 2.0
        total_investment += cost_mid

        # Payback & ROI
        payback_m = int(rec.get("payback_period_months", intervention.get("payback_period_months", 24)))
        weighted_payback_sum += payback_m * cost_mid
        roi_val = float(intervention.get("roi_pct", 25.0))
        roi_list.append(roi_val)

        # CO2 Reduction range midpoint (%)
        co2_range = rec.get("estimated_co2_reduction_range") or [
            intervention.get("estimated_co2_reduction_min", 0),
            intervention.get("estimated_co2_reduction_max", 0)
        ]
        co2_pct_mid = (float(co2_range[0]) + float(co2_range[1])) / 2.0

        # Applicable leak point(s)
        applicable_leaks = intervention.get("applicable_leak_types", [])
        if not applicable_leaks:
            app_leak = rec.get("applicable_leak_point", "general")
            applicable_leaks = [app_leak]

        # Overlap rule: Take max reduction pct for each targeted leak point
        for leak_ref in applicable_leaks:
            leak_key = leak_ref.lower()
            current_max = leak_point_reduction_pcts.get(leak_key, 0.0)
            if co2_pct_mid > current_max:
                leak_point_reduction_pcts[leak_key] = co2_pct_mid

    # Aggregate total non-overlapping reduction percentage (capped at 90%)
    total_reduction_pct = min(90.0, sum(leak_point_reduction_pcts.values()))
    
    reduction_abs = round((current_co2e * (total_reduction_pct / 100.0)), 4)
    projected_co2e = round(max(0.0, current_co2e - reduction_abs), 4)

    average_payback = int(round(weighted_payback_sum / total_investment)) if total_investment > 0 else 0
    avg_roi = round(float(sum(roi_list) / len(roi_list)), 2) if roi_list else 0.0

    return {
        "current_co2e": round(current_co2e, 4),
        "projected_co2e": projected_co2e,
        "reduction_abs": reduction_abs,
        "reduction_pct": round(total_reduction_pct, 2),
        "investment": round(total_investment, 2),
        "payback": average_payback,
        "roi": avg_roi
    }

File: backend/src/services/simulator.js
const Assessment = require('../models/Assessment');
const Recommendation = require('../models/Recommendation');
const Intervention = require('../models/Intervention');

/**
 * What-If Simulator
 * Non-overlapping max-per-leak-point CO2 reduction rule
 * Cost & savings are ADDITIVE across all selected interventions
 */
async function simulate(assessmentId, selectedInterventionIds) {
  const assessment = await Assessment.findById(assessmentId);
  if (!assessment) throw { code: 'NOT_FOUND', message: 'Assessment not found', status: 404 };
  if (assessment.status !== 'complete' || assessment.total_co2e === null) {
    throw { code: 'ASSESSMENT_NOT_CALCULATED', message: 'Assessment not calculated', status: 400 };
  }

  if (!selectedInterventionIds || selectedInterventionIds.length === 0) {
    throw { code: 'VALIDATION_ERROR', message: 'selected_intervention_ids must contain at least 1 item', status: 400 };
  }

  const currentCo2e = assessment.total_co2e;

  // Fetch intervention data for each selected ID
  // IDs can be intervention IDs (INT-xxx) or recommendation IDs (uuid)
  const interventionData = [];
  for (const id of selectedInterventionIds) {
    let intervention = await Intervention.findById(id);
    let applicable_leak_point = null;

    if (!intervention) {
      // Try as recommendation ID
      const rec = await Recommendation.findById(id);
      if (rec) {
        intervention = await Intervention.findById(rec.intervention_id);
        applicable_leak_point = rec.applicable_leak_point;
      }
    }

    if (!intervention) {
      throw { code: 'VALIDATION_ERROR', message: `Unknown intervention ID: ${id}`, status: 400 };
    }

    interventionData.push({ intervention, applicable_leak_point, id });
  }

  // Group by leak_point_ref — apply max-per-leak-point rule for CO2 reduction
  const leakPointGroups = {}; // leak_point_ref -> list of {co2_midpoint, intervention, id}
  const noLeakPointItems = [];

  for (const item of interventionData) {
    const co2Midpoint = (item.intervention.estimated_co2_reduction_min + item.intervention.estimated_co2_reduction_max) / 2;
    const lp = item.applicable_leak_point;

    if (lp) {
      if (!leakPointGroups[lp]) leakPointGroups[lp] = [];
      leakPointGroups[lp].push({ ...item, co2Midpoint });
    } else {
      noLeakPointItems.push({ ...item, co2Midpoint });
    }
  }

  let totalCo2Reduction = 0;
  let totalInvestment = 0;
  let totalAnnualSavings = 0;
  let totalPaybackMonths = 0;
  let paybackCount = 0;
  let totalRoi = 0;
  let roiCount = 0;
  const overlapAdjustments = [];

  // Apply max-per-leak-point rule
  for (const [lpRef, items] of Object.entries(leakPointGroups)) {
    // Sort by co2Midpoint descending — take max
    items.sort((a, b) => b.co2Midpoint - a.co2Midpoint);
    const maxItem = items[0];
    totalCo2Reduction += maxItem.co2Midpoint;

    if (items.length > 1) {
      overlapAdjustments.push({
        leak_point_ref: lpRef,
        applied_intervention_id: maxItem.id,
        note: `max-per-leak-point rule applied; ${items.length - 1} overlapping intervention(s) on this leak point ignored for CO2 reduction, cost still counted`,
      });
    }

    // Cost is ADDITIVE for all items in this group
    for (const item of items) {
      const costMidpoint = (item.intervention.estimated_cost_min + item.intervention.estimated_cost_max) / 2;
      totalInvestment += costMidpoint;
      // Annual savings estimate (cost / payback_months * 12)
      const payback = item.intervention.payback_period_months;
      const annualSavings = costMidpoint / (payback / 12);
      totalAnnualSavings += annualSavings;
      totalPaybackMonths += payback;
      paybackCount++;
      totalRoi += item.intervention.roi_pct;
      roiCount++;
    }
  }

  // Non-grouped items (no leak point) — simple addition
  for (const item of noLeakPointItems) {
    totalCo2Reduction += item.co2Midpoint;
    const costMidpoint = (item.intervention.estimated_cost_min + item.intervention.estimated_cost_max) / 2;
    totalInvestment += costMidpoint;
    const payback = item.intervention.payback_period_months;
    const annualSavings = costMidpoint / (payback / 12);
    totalAnnualSavings += annualSavings;
    totalPaybackMonths += payback;
    paybackCount++;
    totalRoi += item.intervention.roi_pct;
    roiCount++;
  }

  // Cap reduction at 90% of current CO2e
  const maxReduction = currentCo2e * 0.90;
  if (totalCo2Reduction > maxReduction) {
    totalCo2Reduction = maxReduction;
  }

  const projectedCo2e = Math.max(0, currentCo2e - totalCo2Reduction);
  const reductionAbs = currentCo2e - projectedCo2e;
  const reductionPct = currentCo2e > 0 ? reductionAbs / currentCo2e : 0;

  const avgPayback = paybackCount > 0 ? Math.round(totalPaybackMonths / paybackCount) : 0;
  const avgRoi = roiCount > 0 ? parseFloat((totalRoi / roiCount).toFixed(1)) : 0;

  return {
    current_co2e: parseFloat(currentCo2e.toFixed(4)),
    projected_co2e: parseFloat(projectedCo2e.toFixed(4)),
    reduction_abs: parseFloat(reductionAbs.toFixed(4)),
    reduction_pct: parseFloat(reductionPct.toFixed(4)),
    investment: Math.round(totalInvestment),
    currency: 'USD',
    annual_savings: Math.round(totalAnnualSavings),
    payback_period_months: avgPayback,
    roi_pct: avgRoi,
    overlap_adjustments: overlapAdjustments,
  };
}

module.exports = { simulate };
